Add frame statistics UI to with_winit example

armansito opened the pull request:
Added a module for frame time statistics and a UI layer that displays the average, minimum, and maximum frame time alongside FPS. The UI can be toggled by pressing the `S` key.

DJMcNab commented on examples/with_winit/src/lib.rs:
```diff
@@ -249,6 +255,8 @@ fn run(
             let height = render_state.surface.config.height;
             let device_handle = &render_cx.devices[render_state.surface.dev_id];
+            let snapshot = stats.snapshot();
+            let _stats_frame = stats.frame_scope();
```
This is unfortunate, because (at least on my machine), this seems to be vsync locked. On the Tiger demo, I get frame times of nearly exactly `16.62ms`, i.e. slightly less than one vsync interval (other comments cover the gap)

armansito replied:
I agree. In my local benchmarks we were vsync limited (I've been running mostly on 120Hz/90Hz) displays so I don't always hit the limit but vsync does impact the measurement methodology.

Is it possible to configure the winit event loop to not be locked by vsync? Maybe we should use a different (perhaps headless) environment for benchmarks although I really like the interactive set up (I have already identify several adversarial cases by playing with the scene transform alone).

DJMcNab replied:
I know it's possible - we need to set the wgpu present_mode to *something*

I'd rather stay vsyncy by default through. Maybe another toggle?

armansito replied:
Done.

DJMcNab commented on examples/with_winit/src/lib.rs:
```diff
@@ -96,6 +97,8 @@ fn run(
     let mut fragment = SceneFragment::new();
     let mut simple_text = SimpleText::new();
+    let mut stats = stats::Stats::new();
+    let mut stats_toggle = true;
```
Can we rename this to `stats_shown`?

armansito replied:
Done.

DJMcNab commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,169 @@
+impl<'a> Drop for FrameScope<'a> {
+    fn drop(&mut self) {
+        self.stats
+            .add_sample(self.start.elapsed().as_micros() as u64);
```
To ensure that this is measuring time between the same point, rather than rendering time, I think we need to inline the call to `elapsed`, store the new `Instant`, and use that as the start time in `frame_scope`. At the moment, we ignore input processing time, for example, which I don't think is right. 

Actually working out the best measure of FPS has always eluded me.

We also should measure GPU time (e.g. using https://github.com/Wumpf/wgpu-profiler), but wiring that up is non-trivial.

armansito replied:
> We also should measure GPU time (e.g. using https://github.com/Wumpf/wgpu-profiler), but wiring that up is non-trivial.

Oh, the GPU time measurement looks neat. I haven't looked into `wgpu-profiler` very closely but supporting GPU timers out of the box would be great (so far I've had to rely on external tools, such as the Metal GPU Performance monitor). I'll investigate this.

armansito replied:
I made the elapsed time measurement tighter so it should closely match the presentation time. I haven't really had time to investigate wgpu-profiler yet but I plan to spend some time on wiring it up soon.

DJMcNab commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,169 @@
+            self.sum -= oldest;
+        }
+        if micros < self.min {
+            self.min = micros;
```
This could also be written as:
```rust
self.min = self.min.min(micros);
```
Which admittedly does produce semantic satiation. I have no real preference there.

armansito replied:
I like the single line version. Done.

DJMcNab commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,169 @@
+            self.min = micros;
+        }
+        if micros > self.max {
+            self.max = micros;
```
It's unfortunate that this isn't tracked over only the time in `samples`, but instead is global. I think we need a reset keybinding too, maybe `c` for clear?

DJMcNab replied:
That is to say, I'm proposing that we allow <kbd>C</kbd> to reset the timing, and don't concern ourselves with making the minimum being across the set of samples

DJMcNab replied:
That is to say, I'm proposing that we allow <kbd>C</kbd> to reset the extremes, and don't concern ourselves with making the minimum be across the set of samples

armansito replied:
Done.

DJMcNab commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,169 @@
+            offset * Affine::translate((width * 0.67, text_height)),
+            &format!("FPS: {:.2}", self.fps),
+        );
+    }
```
Could we also render a frame time bar graph, showing how stable we are, maybe with some neat colour coding?

armansito replied:
This is absolutely on my TODO list. I'll spend some time this week to incorporate this into this PR (should be easy to add since the code to track samples for the bar graph is already there).

armansito replied:
Done. For now all bars are drawn as green but I can look into some color coding next.

DJMcNab reviewed (commented):
Nothing blocking, some thoughts on potential future work.

DJMcNab commented:
A couple other thoughts; we should record startup time (ideally both time spent in wgpu `create_shader_module` and seperately, time spent in our pre-processing). And we should have a way to open this on Android. What that looks like, I'm not sure

DJMcNab commented on examples/with_winit/src/lib.rs:
```diff
@@ -142,6 +147,23 @@ fn run(
+                                    if vsync_on {
+                                        wgpu::PresentMode::Fifo
+                                    } else {
+                                        wgpu::PresentMode::Immediate
```
```suggestion
                                        wgpu::PresentMode::AutoNoVsync
```
Should be supported on all platforms. There's also a question of doing `PresentMode::AutoVsync` for vsync. I don't have a machine which it would make a difference on; apparently it's only Amd Vulkan.

The alternative might also resolve #266

armansito replied:
Done, I change the code to always use the `Auto*` modes. On web this likely means that "vsync off" will be incorrect so maybe the stats UI shouldn't just blindly display "VSync: off", but it's probably not a huge deal.

DJMcNab commented on examples/with_winit/src/lib.rs:
```diff
@@ -333,6 +366,11 @@ fn run(
             surface_texture.present();
             device_handle.device.poll(wgpu::Maintain::Poll);
+
+            stats.add_sample(stats::Sample {
+                frame_time_us: frame_start_time.elapsed().as_micros() as u64,
+            });
+            frame_start_time = Instant::now();
```
```suggestion
            let new_time = Instant::now();
            stats.add_sample(stats::Sample {
                frame_time_us: (new_time - frame_start_time).as_micros() as u64,
            });
            frame_start_time = new_time;
```
Should be slightly more correct. In practise, it probably doesn't matter

armansito replied:
Done.

DJMcNab reviewed (commented):
Looks good, just a few thoughts on cross-platform support.

I haven't properly reviewed the code for the frame time graph, but it seems to work OK.
We can get overflow in cases of <kbd>C</kbd> being pressed, but that's unavoidable.

I'd also like to get some way to enable this on Android in (without plugging in a USB keyboard, which might just work), but I'm not sure what that would be.

armansito commented:
> Looks good, just a few thoughts on cross-platform support.
> 
> 
> 
> I haven't properly reviewed the code for the frame time graph, but it seems to work OK.
> 
> We can get overflow in cases of <kbd>C</kbd> being pressed, but that's unavoidable.
> 
> 
> 
> I'd also like to get some way to enable this on Android in (without plugging in a USB keyboard, which might just work), but I'm not sure what that would be.

I made the stats display on by default and I can see it on my Android device. Does it show up on your device? 

For controls to hide it I thought of adding a "double tap" gesture or recognizing taps over some visible rect which could serve as a button.

DJMcNab commented:
Ah yes, I hadn't realised the consequences of that; I hadn't tried it on my device. I'm not certain it being displayed by default is what we want, but I don't have a strong objection to it either.

armansito commented:
> Ah yes, I hadn't realised the consequences of that; I hadn't tried it on my device. I'm not certain it being displayed by default is what we want, but I don't have a strong objection to it either.
 
Also there is currently no way to clear min/max on mobile, so we'll want some UI for it. I think can make some simple "double-tap" or some other gesture could work, maybe in another PR.

DJMcNab commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,183 @@
+            // shrink the draw size of the overall average sample, so scale the size non-linearly to
+            // emphasize smaller samples.
+            let h = (*sample as f64) * 0.001 / self.frame_time_max_ms;
+            let s = Affine::scale_non_uniform(1., -h.sqrt());
```
This logic is non-obvious to me; I suspect a maximum height based on the maximum seen last render might be better. I'd much rather keep things linear.

armansito replied:
I switched this to linear now. I thought using sqrt here added value initially but now that we can reset the max/min frame times with a keypress (which resets the scale) this type of visualization is not necessary.

I'll follow up with the dashed lines for 16.6ms/33.3ms as you suggested, which should provide a good relative visual anchor.

DJMcNab reviewed (commented):
Code mostly looks good to me, although I'm not convinced by the non-linearity.

So to confirm,  the current status is:
- https://github.com/linebender/vello/pull/288#issuecomment-1472488546 (mobile buttons) can be done in a different PR
- https://github.com/linebender/vello/pull/288#issuecomment-1458553615 (shader startup time) should be reasonably straightforward, but isn't critical
- https://github.com/linebender/vello/pull/288#discussion_r1138265798 `wgpu_profiler` - doesn't need to be this PR
- https://github.com/linebender/vello/pull/288#discussion_r1138259806 (Colour coding) - would be nice to get in this PR - we should also have dashed lines for (at least) 16.6ms and 33.3 ms. It would also be good to use accessible colours here, rather than red and green, although the colour coding is not functional.

armansito commented:
> Code mostly looks good to me, although I'm not convinced by the non-linearity.
> 
> So to confirm, the current status is:
> 
> * [#288 (comment)](https://github.com/linebender/vello/pull/288#issuecomment-1472488546) (mobile buttons) can be done in a different PR
> * [#288 (comment)](https://github.com/linebender/vello/pull/288#issuecomment-1458553615) (shader startup time) should be reasonably straightforward, but isn't critical
> * [#288 (comment)](https://github.com/linebender/vello/pull/288#discussion_r1138265798) `wgpu_profiler` - doesn't need to be this PR

Working on these in follow up PRs sounds good to me.

> * [#288 (comment)](https://github.com/linebender/vello/pull/288#discussion_r1138259806) (Colour coding) - would be nice to get in this PR - we should also have dashed lines for (at least) 16.6ms and 33.3 ms. It would also be good to use accessible colours here, rather than red and green, although the colour coding is not functional.

I added color coding and lines and picked accessible colors for the coding. I added markers for 8.33, 16.66, and 30.33. I used a solid stroke instead of a dashed stroke as the dashed stroke style didn't work.

DJMcNab reviewed (approved):
I'm happy for this to land as is. I've added some more thoughts on Zulip about the scale, so they might be worth considering. We might discuss tomorrow

armansito commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,221 @@
+                    left_margin_padding,
+                    (1 + labels.len()) as f64 * text_height,
+                )) * s,
+                if *sample < 16667 {
```
Done.

armansito commented on examples/with_winit/src/stats.rs:
```diff
@@ -0,0 +1,221 @@
+                        left_margin,
+                        (2. - y) * graph_max_height + thres_text_height_2,
+                    )),
+                &format!("{}", t),
```
I think that could be neat, though it's a minor visual improvement given how small the label is rendered. @dfrg Is this straightforward to do this today with the existing code?

dfrg replied:
We don’t currently do any sort of alignment but we could probably hack something in.
